ambari-server/src/main/java/org/apache/ambari/server/controller/internal/UpgradeResourceProvider.java

```diff
@@ -755,7 +756,17 @@ major stack versions (e.g., HDP 2.2 -> 2.3), and then set config changes
     at the appropriate moment during the orchestration.
     */
     if (pack.getType() == UpgradeType.ROLLING || pack.getType() == UpgradeType.HOST_ORDERED) {
+      if (direction == Direction.UPGRADE) {
+        StackEntity targetStack = upgradeContext.getRepositoryVersion().getStack();
+        cluster.setDesiredStackVersion(
+            new StackId(targetStack.getStackName(), targetStack.getStackVersion()));
+      }
       s_upgradeHelper.updateDesiredRepositoriesAndConfigs(upgradeContext);
+      if (direction == Direction.DOWNGRADE) {
+        StackId targetStack = upgradeContext.getCluster().getCurrentStackVersion();
+        cluster.setDesiredStackVersion(
```

Comment:
nit: you already have a StackId, no need to make a new one.

Reply:
Also, did you happen to test an EU?

Reply:
Tested EU, works well